ggml/src/ggml-cpu/ggml-cpu-aarch64.c

```diff
@@ -1090,466 +1090,62 @@ void ggml_gemm_q4_0_4x4_q8_0(int n, float * restrict s, size_t bs, const void *
-            "cmp x10, #0x1\n"
-            "str q15, [x20, #0x0]\n"
-            "add x20, x20, %x[res_stride]\n"
-            "ble 8f\n"
-            "cmp x10, #0x2\n"
-            "str q19, [x20, #0x0]\n"
-            "add x20, x20, %x[res_stride]\n"
-            "ble 8f\n"
-            "cmp x10, #0x3\n"
-            "str q18, [x20, #0x0]\n"
-            "add x20, x20, %x[res_stride]\n"
-            "ble 8f\n"
-            "str q14, [x20, #0x0]\n"
-            "8:"  // Row tail: Accumulator store skip
-            "subs x23, x23, #0x4\n"
-            "add %x[res_ptr], %x[res_ptr], #0x10\n"
-            "bne 6b\n"
-            "subs x10, x10, #0x4\n"
-            "add %x[a_ptr], %x[a_ptr], x9\n"
-            "mov %x[res_ptr], x22\n"
-            "bgt 5b\n"
-            "9:"  // Row tail: Row loop skip
-            : [a_ptr] "+&r" (a_ptr), [res_ptr] "+&r" (res_ptr)
-            : [b_ptr] "r" (b_ptr), [nr] "r" (nr), [nb] "r" (nb), [res_stride] "r" (res_stride), [nc] "r" (nc)
-            : "cc", "memory", "v0", "v1", "v2", "v3", "v4", "v5", "v6", "v7", "v8", "v9", "v10", "v11", "v12", "v13", "v14", "v15", "v16", "v17", "v18", "v19", "v20", "v21", "v22", "v23", "v24", "v25", "v26", "v27", "v28", "v29", "v30", "v31", "x9", "x10", "x20", "x21", "x22", "x23", "x24", "x25", "x26", "x27", "x28"
-        );
-        return;
+                    for (int k = 0; k < 4; k++) {
+                        int8x16_t b0 = vld1q_s8((const int8_t *)b_ptr->qs + 16 * k);
+                        int8x16_t b1 = b0 & 0xF0U;
+                        b0 <<= 4;
+
+                        int8x16_t a0 = vld1q_s8(a_ptr->qs + 16 * k);
+                        int8x16_t a1 = vld1q_s8(a_ptr->qs + 16 * k + 4 * qk/2);
+
+                        ret0 = vdotq_laneq_s32(ret0, b0, a0, 0);
+                        ret1 = vdotq_laneq_s32(ret1, b0, a0, 1);
+                        ret2 = vdotq_laneq_s32(ret2, b0, a0, 2);
+                        ret3 = vdotq_laneq_s32(ret3, b0, a0, 3);
+
+                        ret0 = vdotq_laneq_s32(ret0, b1, a1, 0);
+                        ret1 = vdotq_laneq_s32(ret1, b1, a1, 1);
+                        ret2 = vdotq_laneq_s32(ret2, b1, a1, 2);
+                        ret3 = vdotq_laneq_s32(ret3, b1, a1, 3);
+                    }
+                    acc0 = vfmaq_f32(acc0, vcvtq_n_f32_s32(ret0, 4), vmulq_laneq_f32(bd, ad, 0));
+                    acc1 = vfmaq_f32(acc1, vcvtq_n_f32_s32(ret1, 4), vmulq_laneq_f32(bd, ad, 1));
+                    acc2 = vfmaq_f32(acc2, vcvtq_n_f32_s32(ret2, 4), vmulq_laneq_f32(bd, ad, 2));
+                    acc3 = vfmaq_f32(acc3, vcvtq_n_f32_s32(ret3, 4), vmulq_laneq_f32(bd, ad, 3));
+                    a_ptr++;
+                    b_ptr++;
+                }
+                vst1q_f32(s + (r + 0) * bs + c, acc0);
+                vst1q_f32(s + (r + 1) * bs + c, acc1);
+                vst1q_f32(s + (r + 2) * bs + c, acc2);
+                vst1q_f32(s + (r + 3) * bs + c, acc3);
+            }
+        }
```

Comment:
```suggestion
        }
        
        return;
```

Reply:
Thanks! Was looking for the huge perf regression...